### PR TITLE
Fix add overlay image over a gif image

### PR DESCRIPTION
**Pull request by mhgtlc**
## Changes in this pull request  
Resolves #
When applying transformation: `Add Overlay (Imagick)` over a gif image, it is applied only to the first frame:
Screenshot 2024-08-01 at 12 56 44

Wrong:
wrong

Right:
right

### CODE REVIEW

**Comment by mhgtlc** [comment]
@brusch please take a look, pipeline failed, but I just copied a sample of code from the same class and apply the same logic for gif as there are in other places in the class

**Comment by brusch** [comment]
@mhgtlc thanks for your PR. Please have a look here: 
https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&pullRequest=17434&types=CODE_SMELL&inNewCodePeriod=true&sinceLeakPeriod=true&id=pimcore_pimcore 

Should be easy to fix 😊 
Thanks a lot

File: lib/Image/Adapter/Imagick.php [hunk 1]
```diff
@@ -744,28 +744,57 @@ public function addOverlay(mixed $image, int $x = 0, int $y = 0, int $alpha = 10
...
+        if ($newImage instanceof \Imagick) {
+            [$x, $y] = $this->calculateOverlayPosition($newImage, $x, $y, $origin);
 
             $newImage->evaluateImage(\Imagick::EVALUATE_MULTIPLY, $alpha, \Imagick::CHANNEL_ALPHA);
-            $this->resource->compositeImage($newImage, constant('Imagick::' . $composite), (int)$x, (int)$y);
+
+            $compositeValue = constant('Imagick::' . $composite);
+            if ($this->checkPreserveAnimation()) {
+                foreach ($this->resource as $frame) {
+                    $frame->compositeImage($newImage, $compositeValue, (int)$x, (int)$y);
```
Comment (blankse) [inline lib/Image/Adapter/Imagick.php]:
```suggestion
                    $frame->compositeImage($newImage, $compositeValue, $x, $y);
```

File: lib/Image/Adapter/Imagick.php [hunk 2]
```diff
@@ -744,28 +744,57 @@ public function addOverlay(mixed $image, int $x = 0, int $y = 0, int $alpha = 10
...
+                }
+            } else {
+                $this->resource->compositeImage($newImage, $compositeValue, (int)$x, (int)$y);
```
Comment (blankse) [inline lib/Image/Adapter/Imagick.php]:
```suggestion
                $this->resource->compositeImage($newImage, $compositeValue, $x, $y);
```

File: lib/Image/Adapter/Imagick.php [hunk 3]
```diff
@@ -744,28 +744,57 @@ public function addOverlay(mixed $image, int $x = 0, int $y = 0, int $alpha = 10
...
+            }
         }
 
         $this->postModify();
 
         return $this;
     }
 
+    public function calculateOverlayPosition(\Imagick $newImage, int $x, int $y, string $origin): array
```
Comment (blankse) [inline lib/Image/Adapter/Imagick.php]:
```suggestion
    /**
     * @return array{int, int}
     */
    private function calculateOverlayPosition(\Imagick $newImage, int $x, int $y, string $origin): array
```

File: lib/Image/Adapter/Imagick.php [hunk 4]
```diff
@@ -744,28 +744,57 @@ public function addOverlay(mixed $image, int $x = 0, int $y = 0, int $alpha = 10
...
+
+                break;
+            case 'center':
+                $x = round($imageWidth / 2 - $newImageWidth / 2) + $x;
+                $y = round($imageHeight / 2 - $newImageHeight / 2) + $y;
+
+                break;
+        }
+
+        return [$x, $y];
```
Comment (blankse) [inline lib/Image/Adapter/Imagick.php]:
```suggestion
        return [(int)$x, (int)$y];
```

**Review by blankse — COMMENTED** [review]
I think the new method should be private

File: lib/Image/Adapter/Imagick.php [hunk 5]
```diff
@@ -985,6 +983,40 @@ public function supportsFormat(string $format, bool $force = false): bool
...
+
+                break;
+            case 'center':
+                $x = round($imageWidth / 2 - $newImageWidth / 2) + $x;
+                $y = round($imageHeight / 2 - $newImageHeight / 2) + $y;
+
+                break;
+        }
+
+        return [(int)$x, (int)$y];
```
Comment (jdreesen) [inline lib/Image/Adapter/Imagick.php]: What do you think of this?

```suggestion
        return match ($origin) {
            'top-right' => [
                $imageWidth - $newImageWidth - $x,
                $y,
            ],
            'bottom-left' => [
                $x,
                $imageHeight - $newImageHeight - $y,
            ],
            'bottom-right' => [
                $imageWidth - $newImageWidth - $x,
                $imageHeight - $newImageHeight - $y,
            ],
            'center' => [
                (int) round($imageWidth / 2 - $newImageWidth / 2) + $x,
                (int) round($imageHeight / 2 - $newImageHeight / 2) + $y,
            ],
            default => [
                $x,
                $y,
            ],
        };
```

**Comment by lukmzig** [comment]
@mhgtlc thank you very much for the fix!

**Pull request merged**